src/provider/model/EventPump.ts

```diff
@@ -0,0 +1,67 @@
+import {injectable, inject} from 'inversify';
+import {Identity} from 'openfin/_v2/main';
+
+import {Inject} from '../common/Injectables';
+import {APITopic, Events} from '../../client/internal';
+import {Transport, Targeted} from '../../client/EventRouter';
+import {NotificationActionEvent} from '../../client';
+
+import {APIHandler} from './APIHandler';
+import {ClientHandler} from './ClientHandler';
+
+type DeferrableEvent = Targeted<Transport<NotificationActionEvent>>;
+type DeferredEvent = {target: Identity, event: DeferrableEvent};
+
+/**
+ * Notification event message handler. EventPump internally handles the events added
+ * by either dispatching them immediately or deferring the dispatch.
+ *
+ */
+@injectable()
+export class EventPump {
+    @inject(Inject.API_HANDLER)
+    private _apiHandler!: APIHandler<APITopic, Events>;
+
+    private _clientHandler!: ClientHandler;
+
+    private _deferredEvents: DeferredEvent[] = [];
+
+    constructor(@inject(Inject.CLIENT_HANDLER) clientHandler: ClientHandler) {
+        this._clientHandler = clientHandler;
+        this._clientHandler.onClientHandshake.add(this.dispatchDeferredEvents, this);
+    }
+
+    /**
+     * Adds an event to EventPump to be dispatched to the specified client.
+     * Type args:
+     *   T: Defines the type of event which occured and needs to be passed to the client.
+     *
+     * @param target Identity of target client.
+     * @param event Notification event to be dispatched
+     *
+     */
+    public async push<T extends Events>(target: Identity, event: Targeted<Transport<T>>): Promise<void> {
+        if (event.type !== 'notification-action' || this._clientHandler.isAppActive(target)) {
+            this._apiHandler.dispatchEvent(target, event);
+        } else {
+            this._deferredEvents.push({target, event: event as DeferrableEvent});
+            this._clientHandler.tryLaunchApplication(target);
```

Comment:
Is there an edge case here where we could try launching the same app twice?

When we have `onClose` actions, this could be called several times in the same 'frame'. There are API's and buttons in the UI that can close multiple notifications simultaneously, all of which could potentially try starting the application.

Reply:
I think the most granular I can get is to check if the app is already running before start call unless we want to stick **flags** to synchronise these further (I know I shouldn't have used the **f-word** :P)
I think the cases like you explained should be handled at a higher level case by case. e.g. a bulk close can be handled as a single action and the EventPump can be referred only once in such actions?

I am open for a discussion and suggestions about this one.

Reply:
oh btw now with the exception handling code in, the start call awaits. this might automatically resolve the issue?

Reply:
I'm getting flashbacks... :-)

> the start call awaits. this might automatically resolve the issue?

I'm not sure this would, since whilst there's `await`s within the app launch code, there's nothing to  ensure that app launches happen sequentially (if there's more than one happening at once - like the example above). The promise also isn't propagated to the `tryLaunchApplication` function itself, meaning that the calling code couldn't handle this queuing either.

> f-word

My main objection to flags (which I'm glad you've taken to heart) is when they bloat API's or duplicate state that exists elsewhere. If this is private state within the component that is responsible for restarting apps, then that's good with me. I think there's a bit of a delay between calling `startFromManifest`, and the app appearing in a `getAllApplications` or `isRunning` check - so relying on the API as "source of truth" here may introduce some edge cases...